Mathlib/Tactic/Basic.lean

```diff
@@ -188,10 +188,49 @@ macro_rules
 
 macro (name := «sorry») "sorry" : tactic => `(exact sorry)
 
--- TODO `n:num` should be an optional argument; if missing, repeat until failure
-elab "iterate " n:num seq:tacticSeq : tactic => do
-  for i in [:n.toNat] do
-    evalTactic seq
+/--
+`iterate_at_most n t` runs the tactic `t` at most `n` times, or until failure,
+whichever comes first.
+Always succeeds, and returns at list of the succesful results.
+-/
+def iterate_at_most : Nat → TacticM α → TacticM (List α)
```

Comment:
I would like to change all the `TacticM`s here to `MetaM`. This works until we get to the porcelain tactic. How do I forget that `evalTactic seq` is a `TacticM` and just remember it as a `MetaM`, so I can pass it to the plumbing?

Reply:
I would generalize `TacticM` to `m` with `[Monad m]` and `[MonadExcept e m]`.